build/xtask/src/lsp.rs

```diff
@@ -0,0 +1,390 @@
+                        (cra, Some(fea))
+                    } else {
+                        (s, None)
+                    };
+
+                    changed |= package_dependencies
+                        .entry(pkg_name)
+                        .or_default()
+                        .insert(cra.to_owned());
+
+                    let d = pkg
+                        .dependencies
+                        .iter()
+                        .find(|d| d.name == cra)
+                        .unwrap();
+                    if d.kind != DependencyKind::Build {
+                        todo.push((cra.to_owned(), fea.map(|s| s.to_owned())))
+                    }
+                }
+            }
+            if !changed {
+                break;
+            }
+            // Start the loop anew, checking whether the optional `crate?/feat`
+            // are now active.
+            assert!(todo.is_empty());
+            todo = optional;
+        }
+
+        package_features
+    }
+}
+
+/// Checks whether the given package is valid for the given task
+fn check_task(
+    package_name: &str,
+    task_name: &str,
+    app_name: &str,
+    app_cfg: &PackageConfig,
+    packages: &PackageGraph,
+) -> Option<LspConfig> {
+    let task = &app_cfg.toml.tasks[task_name];
+
+    // Check to see if our target package is used in this task (resolved based
+    // on per-task features)
+    let dependencies = packages.resolve(&task.name, &task.features);
+
+    // Congrats, we've found a task in the given image which uses our
+    // desired crate.  Let's do some stuff with it.
+    if dependencies.contains_key(package_name) {
+        let build_cfg = app_cfg
+            .toml
+            .task_build_config(&task_name, false, None)
+            .map_err(|_| {
+                anyhow!("could not get build config for {}", task_name)
+            })
+            .unwrap();
+
+        let mut iter = build_cfg.args.iter();
+        let mut target = None;
+        while let Some(t) = iter.next() {
+            if t == "--target" {
+                target = iter.next().to_owned();
+            }
+        }
+
+        if target.is_none() {
+            panic!("missing --target argument in build config");
+        }
+
+        let features: Vec<String> = dependencies
+            .iter()
+            .flat_map(|(package, feat)| {
+                feat.iter().map(move |f| format!("{package}/{f}"))
+            })
+            .collect();
+
+        let mut build_override_command: Vec<String> =
+            "cargo check --message-format=json"
+                .split(' ')
+                .map(|s| s.to_string())
+                .collect();
+        build_override_command.extend(build_cfg.args.iter().cloned());
+        build_override_command
+            .extend(dependencies.keys().map(|p| format!("-p{p}")));
+        build_override_command.push("--target-dir".to_owned());
+        build_override_command.push("target/rust-analyzer".to_owned());
```

Comment:
I also use this target dir for rust-analyzer, but is this personal-config-specific?

Reply:
On second thought, it's useful, but let's do it on the editor side (to make it more flexible); done in bfb8e1ed52e934cd0038ac8a4e8967d9e1b24a35